### PR TITLE
[FLINK-21086][runtime][checkpoint] CheckpointBarrierHandler Insert barriers for channels received EndOfPartition

### PR DESCRIPTION
**Pull request by gaoyunhaii**
## What is the purpose of the change

This PR makes `CheckpointBarrierHandler` inserts barrier for channels that have received EndPartition but not processed it yet. 

When an input channel received EndOfPartition, it would insert a prioritized `FinalizeBarrier` at the head of queue, which indicating to `CheckpointBarrierHandler` that this channel has received `EndOfPartition` and what its next expected barrier id is. 

For `CheckpointBarrierHandler`, it introduce a new component `FinalizeBarrierComplementProcessor` to deal with `FinalizedBarrier` and insert barriers.
1.  When start checkpoint due to received barrier from another channel or RPC trigger, it would insert barriers for all the input channels received EndOfPartition
2. When received `FinalizeBarrier` from one input channel, it would insert barriers for all the pending checkpoints for this channel.
It would also record the next expected barrier id for each channel and would not insert stale barriers.

**LocalInputChannel** 

Previously LocalInputChannel would directly pull buffer from `ResultPartition` and would not cache buffers. However, with inserted barriers, LocalInputChannel would have to support caching some buffers. We would like to use a FIFO queue to cache buffers instead of a prioritized queue to simplify the notification to InputGates (e.g., do not need to notifyPrioritizedEvent for the `LocalInputChannel` as before).

## Brief change log

- 7b5bcb2aea349560a6fac8d07e879f6e135f2e67 introduces the `FinalizedBarrier` event.
- 6f78a9a3c18c8ed8c43cfe6213e80c793e84f5b3 InputChannels insert `FinalizedBarrier` before EndOfPartition and support inserting barriers before EndOfPartition
- 1240dbb9879bedee17ac69d69ab8fd514589febd Modify the `CheckpointBarrierHandler` to support insert barriers when needed and support alignment with EndOfPartition.

## Verifying this change

This change added tests and can be verified via added UT for input channels and checkpoint barrier handlers.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): **no**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: **no**
  - The serializers: **no**
  - The runtime per-record code paths (performance sensitive): **no**
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: **no**
  - The S3 file system connector: **no**

## Documentation

  - Does this pull request introduce a new feature? **no**
  - If yes, how is the feature documented? **not applicable**

### CODE REVIEW

**Review by pnowojski — CHANGES_REQUESTED** [review]
Thanks @gaoyunhaii, I've left couple of inlined comments, however I have a couple of general concerns.

First, it seems like unaligned checkpoints are not working properly with the `RemoteInputChannel`, if there are unprocessed buffers still enqueued in the upstream's `PiepelinedSubpartition`. In your implementing it looks those buffers can not be overtaken and unaligned checkpoints start working again only once `EndOfPartitionEvent` is received by the `RemoteInputChannel` on the downstream node?

Secondly wouldn’t it be simpler to keep the task alive until its partitions are consumed? It could enqueue `EndOfPartitionEvent` as it’s doing normally. If it receives trigger aligned checkpoint RPC, it could completely ignore it (`EndOfPartitionEvent` could trigger aligned checkpoint downsteam). If it receives unaligned checkpoint  request, it could follow the normal procedure.

1. Overtake output in-flight buffers and persist them on state (note, at this point we might have skipped checkpointing state of the operators, as they are already closed)
2. Enqueue unaligned checkpoint barrier at the head of the `PipelinedSubpartition`'s queue.
3. Overtake input in-flight buffers down stream, and process the checkpoint completely normally.

As far as I can tell, it would help us avoid all of the network stack changes, except of special handling `EndOfPartitionEvent` (instead of failing checkpoints, it would decrease the number of "open" channels).

Furthermore, I remember there were some orthogonal discussions with @tillrohrmann that 
general it would be better to keep `StreamTask` alive until its partition is fully consumed. I don't remember now the full context, but part of the problem was resource management. Currently `StreamTask` would claim that it has released its resources and its task slot, while in reality some of those resources are still being used (for example memory buffers).

PS. For the first version, we could first ignore unaligned checkpoints. We could still target unaligned checkpoints support for 1.13, but we could split it into a separate ticket.

**Comment by gaoyunhaii** [comment]
Hi @pnowojski , very thanks for the review and deep thoughts! I first tried to have a consideration for the general concerns.

For the first issue, previously Roman and I also have some discussion, we think that we might solve this issue via making the upstream tasks to wait till all the buffers are flushed out of the sub-partitions before get to finished. Then during this period, the upstream task would be still running and we could then do unaligned checkpoint normally.

For holding the upstream tasks till all the remaining data is consumed, I also think it is a good direction that might simplify the issue. But I still found it might not easy to implement in some detailed points. For convenience suppose we have one task B with two upstream tasks A1 and A2.

The first concern is about the upstream tasks ignores the barriers to the downstream tasks after `EndOfPartition` event in the aligned checkpoint case:

1. If both A1 and A2 received barrier after EndOfPartition, then task B would only receive EndOfPartitions and would not know any thing about this checkpoint. JM also could not detect this situation since all the checkpoint RPC triggers are success. This would cause checkpoint timeout and then cause job failover. 


The second concern is about letting the barriers to "chase after" EndOfPartition for unaligned checkpoint case:

1. Since EndOfPartition is sent in `Task` instead of `StreamTask`, when EndOfPartition is send, the `StreamTask` has finished its lifecycle, and the `SubtaskCheckpointCoordinator` is also closed in `StreamTask#cleanupInvoke()`. This makes us could not snapshot the result partition since `ChannelStateWriter` is closed.
2. The issue for aligned checkpoint still exists: if both A1 and A2 received barrier after EndOfPartition, the barriers might not overtake EndOfPartition and B will have no idea about this checkpoint.
3. The aligned with timeout checkpoint is not still supported, since in this case the barrier is not prioritized and cannot go past EndOfPartition.
4. One potential issue is that we send some events after EndOfPartition might break some assumptions for the network stack. We would need to add a check along the sub-partition and input channel to deal with the case that the barrier arrives after the channel is already closed due to EndOfPartition. 


There are also some implementation related issues, these issues should be able to be solved, but might introduce some complexity:

1. We would need to introduce a new event for the downstream task to notify the upstream task to exit. We could not reuse CloseRequest since it is connection level and there might be multiple tasks share the same connection. 
2. Besides, the upstream task need to do a reference count about how much exit events are received from the downstream tasks. 
3. Since the EndOfPartition event is send in `Task`, we would also need some method to avoid affect `BatchTask`. We might need to add some mark interface to AbstractInvokable.
4. We would still need to do some modification to the downstream tasks: support A1 is running and A2 has finished, then when B received unaligned checkpoints from A1, it might decide to trigger the checkpoint and snapshot the buffers for A2. Thus `RemoteInputChannel` would still need to modify the code the consider EndOfPartition in channel persistence logic.

**Comment by pnowojski** [comment]
Thanks for the explanation @gaoyunhaii . I've thought about what you had written, and I still have some concerns.

> If both A1 and A2 received barrier after EndOfPartition, then task B would only receive EndOfPartitions and would not know any thing about this checkpoint. JM also could not detect this situation since all the checkpoint RPC triggers are success. This would cause checkpoint timeout and then cause job failover.

Why RPC triggers would need to be successful? We can modify the logic in `StreamTask`, to fail them as we wish (more about this later).

> The second concern is about letting the barriers to "chase after" EndOfPartition for unaligned checkpoint case

But here you are not solving this problem. Unaligned Checkpoints in your proposal are not able to overtake the remaining output buffers. You are adding quite a bit more complexity and special handling on the input side to partially support unaligned checkpoints, while in the long term we would still need to keep the upstream task alive. Your solution doesn't bring us closer to that, while it adds extra complexity that won't be needed in the final version, right?

For me, modifying the closing protocol, to add some kind of hand shake, would be more generic and cleaner solution. It would not affect alignment/unaligned checkpoints, apart from one minor thing: we might need to persist some events (like `EndOfPartitionEvent`) if they are overtaken by the unaligned `CheckpointBarrier`. As I said at the beginning: adding code/complexity to `StreamTask`/`Task` or even to closing protocol is much much safer compare to altering the behaviour of the alignment (caching logic in `LocalnputChannel` or `#insertBarrierBeforeEndOfPartition`).

As I see, the alternate approach would be something like that:

1. introduce a two step handshake when closing channels. First upstream sends `EndOfPartitionEvent`, downstream upon processing it sends `PartitionConsumedEvent`, upstream sends `CloseEvent`
2. downstream should expect some events, like `CheckpointBarrier` to be sent after the `EndOfPartitionEvent`. `CloseEvent` would let the downstream know it can safely release all of the resources.
3. After sending `EndOfPartitionEvent` upstream would be in `FINISHING` state. It can still handle checkpoint RPCs.
4. Upstream after sending `CloseEvent` would switch to `FINISHED` state and reject all checkpoint RPCs. 
5. If checkpoint trigger RPC is rejected because task switched to `FINISHED`, `CheckpointCoordinator` should take that into account and redirect the RPC to all tasks downstream from the `FINISHED` one. I assume you need this behaviour also in the current version? @gaoyunhaii what are you doing if there is a race condition and some `Task` finishes while trigger checkpoint RPC is already on its way to it?


> Besides, the upstream task need to do a reference count about how much exit events are received from the downstream tasks

We are already doing that (`PipelinedResultPartition#numUnconsumedSubpartitions`).

**Comment by gaoyunhaii** [comment]
Hi @pnowojski , very thanks for the deep thoughts! I also like the idea of let the upstream tasks and downstream tasks to do handshake before exit, and I only have some minor complements:
1. I also agree with we could do a two step handshake. Since we now already have `EndOfpartitionEvent` as the last event, I think we could introduce a new event before `EndOfpartitionEvent` like `EndOfInputEvent` (or some other names) ? Namely change `EndOfpartitionEvent` to `EndOfInputEvent` and make `CloseEvent` to `EndOfPartitionEvent`, and we could also deal with all the logic in `StreamTask`. 
2. If a task received RPC trigger, it indicates it must have processed all the `EndOfInputEvent` and might only have some pending CheckpointBarrier to process. In this case the task could bookkeep the received RPC checkpoint triggers, and after all the EndOfPartition is processed, it could then trigger all the bookkept checkpoints. 

3. > I assume you need this behaviour also in the current version? @gaoyunhaii what are you doing if there is a race condition and some Task finishes while trigger checkpoint RPC is already on its way to it?

Yes, indeed, if a task finished before it received the trigger message, in the first version we would directly decline the checkpoint ([FLINK-21246](https://issues.apache.org/jira/browse/FLINK-21246)), and for more fine-grained processing we could recompute and re-trigger the following tasks ([FLINK-21081](https://issues.apache.org/jira/browse/FLINK-21081)).

I'll open a new PR based on this solution.

**Comment by pnowojski** [comment]
Could you rephrase what do you mean by:
> If a task received RPC trigger, it indicates it must have processed all the EndOfInputEvent and might only have some pending CheckpointBarrier to process. In this case the task could bookkeep the received RPC checkpoint triggers, and after all the EndOfPartition is processed, it could then trigger all the bookkept checkpoints.

?

**Comment by gaoyunhaii** [comment]
Suppose a task `B` has two precessors `A1` and `A2`, `A1` and `A2` has sent `Barrier_5` and `Barrier_6` after `EndOfInputEvent` and then finish. The two barrier might just reached `B`'s input channels and are not processed yet. At the same time, the `CheckpointCoordinator` starts a new checkpoint and since `A1` and `A2` are finished, `B` becomes the new source and `CheckpointCoordinator` would trigger `Barrier_7` to `B`. In this case, `Barrier_7` could be viewed as a new checkpoint come after all the un-processed ones (namely `Barrier_5` and `Barrier_6`), thus we could trigger checkpoint 7 after all the un-process ones, namely after we processed `Barrier_5` and `Barrier_6`.

**Comment by pnowojski** [comment]
I guess you are right. I agree that in that case probably the easiest thing to do, would be to just skip `Barrier_5 and Barrier_6` and  subsume the older barriers with the `Barrier_7`.

**Comment by gaoyunhaii** [comment]
Yes, subsume would indeed be easier, the only concern is that it might cause some checkpoint failures that might introduce some confusion to users. Do you think this would be an issue or we could just ignore it and use subsume~?

**Comment by pnowojski** [comment]
Subsuming is fine. It can happen right now just as well pretty frequently (especially with part of the job heavy backpressured).

**Comment by gaoyunhaii** [comment]
Got it~ I'll go with the subsume~

**Comment by kezhuw** [comment]
@gaoyunhaii @pnowojski I have some immature and unverified thoughts.

I think handling of `EndOfPartitionEvent` during checkpoint and overtaking buffers before `EndOfPartitionEvent` are kind of orthogonal and could be coexist. We will encounter `EndOfPartitionEvent` during checkpoint regardless, canceling checkpoint in this case should be avoided after FLIP-147.

But instead of "Insert barriers for channels received EndOfPartition", I would suggest to count `EndOfPartitionEvent` directly in checkpoint barrier handler. That is, in implementation, we could count/trigger checkpoint in `CheckpointBarrierHandler.processEndOfPartition` and `ChannelStatePersister.checkForBarrier` if there is pending checkpoint. Personally, I think it is same as `FinalizeBarrierComplementPrcoessor` but much straightforward. The point is if we decide to support `EndOfPartitionEvent` during checkpoint, we should burn this knowledge directly to `CheckpointBarrierHandler`.

To overtake buffers before `EndOfPartitionEvent` for unaligned checkpoint, I think the minimal requirement is flushing all buffers to network stack before `EndOfPartitionEvent`. An asynchronous completable flush operation on `PipelinedSubpartition` should meet this requirement. Before that flush operation completed, unaligned checkpoint could take place as normal, after that there will be no output buffers to overtake. Also, an request-response paired events(eg. `EndOfXyzEvent`, `XyzConsumedEvent`) will fulfill this requirement. I am not sure how it is viable to introduce `FINISHING` for checkpointing as I think `ExecutionState` is tackled by `Task` while checkpointing is tackled inside `StreamTask`. But I think this "how to overtake buffers for unaligned checkpoint before `EndOfPartitionEvent`" could be a separated issue.

Besides this, I think checkpoint after `EndOfInputEvent`(eg. at `FINISHING` or all buffer flushed) is similar to last checkpoint for 2pc. I am kind of worry about buffer duplication after recovery a successful checkpoint created in that period. Checkpoint in that period will include last buffer from possible end-of-stream-flush operation, after recovery, that end-of-stream-flush will still be executed ?

I saw there are other jiras in this pr or planed pr, I want to writes my thoughts also:
* FLINK-21085: I think the currently solution will undermine what we try to follow in FLINK-21133. In that jira, we try to unify handling of stop-with-savepoint in one place, while the currently approach tends to duplicate checkpoint trigger in many code paths. I would like to suggest to enhance and generalize `MultipleInputStreamTask.triggerCheckpointAsync` to `StreamTask`.  This way we will have only place to trigger checkpoint.
* FLINK-21081:  Just yet another approach to list/consider/evaluation. May be we could send checkpoint trigger rpc to all running tasks ? If tasks with active inputs receive checkpoint trigger rpc, just bookkeeping it. If all active inputs are `EndOfPartitionEvent` in the meantime, task will trigger checkpoint themselves. If checkcpoint-trigger is coded in one place, this bookkeeping and lazy-trigger will only be coded once in one place. But this approach may be network-consuming in large job topology.

@pnowojski I am no sure how finishing handshake will be delivered, so basically, I am also not sure how this handshaking will combine with stop-with-savepoint. But if the handshaking happens at `StreamTask` level, I think there will be no big deal. Though, still unsure.

**Comment by kezhuw** [comment]
@gaoyunhaii @pnowojski I want to describe an example for necessity of `EndOfPartitionEvent` tolerance during checkpoint.

Assumes that operator `C` has two upstream `A` and `B`. During a checkpoint, `A` ends with `EndOfPartitionEvent` while `B` keeps running. `C` could receive checkpoint barrier from `B` and then `EndOfPartitionEvent` from `A`. In this case, FLINK-21081 does not help.

**Comment by gaoyunhaii** [comment]
Hi @kezhuw very thanks for the deep thoughts and suggestions! 

> But instead of "Insert barriers for channels received EndOfPartition", I would suggest to count EndOfPartitionEvent directly in checkpoint barrier handler. That is, in implementation, we could count/trigger checkpoint in CheckpointBarrierHandler.processEndOfPartition and ChannelStatePersister.checkForBarrier if there is pending checkpoint. Personally, I think it is same as FinalizeBarrierComplementPrcoessor but much straightforward. The point is if we decide to support EndOfPartitionEvent during checkpoint, we should burn this knowledge directly to CheckpointBarrierHandler.
> 
> To overtake buffers before EndOfPartitionEvent for unaligned checkpoint, I think the minimal requirement is flushing all buffers to network stack before EndOfPartitionEvent. An asynchronous completable flush operation on PipelinedSubpartition should meet this requirement. Before that flush operation completed, unaligned checkpoint could take place as normal, after that there will be no output buffers to overtake. Also, an request-response paired events(eg. EndOfXyzEvent, XyzConsumedEvent) will fulfill this requirement. I am not sure how it is viable to introduce FINISHING for checkpointing as I think ExecutionState is tackled by Task while checkpointing is tackled inside StreamTask. But I think this "how to overtake buffers for unaligned checkpoint before EndOfPartitionEvent" could be a separated issue.

Yes, as the discussion with Piotr in the above, I think the listed process here is similar to the new process we proposed. One difference is that now we wait not only until all data have been flushed, but also the data has been received and processed by the downstream tasks. After that all the user records are processed, then the upstream task could emit EndOfPartition, and the `CheckpointBarrierHandler` in the downstream task would be able to view `EndOfPartition` as a sign that all the pending checkpoints should be finished.

> Besides this, I think checkpoint after EndOfInputEvent(eg. at FINISHING or all buffer flushed) is similar to last checkpoint for 2pc. I am kind of worry about buffer duplication after recovery a successful checkpoint created in that period. Checkpoint in that period will include last buffer from possible end-of-stream-flush operation, after recovery, that end-of-stream-flush will still be executed ?

No, the waiting happens after all the operators have been closed, and the operator subtask would be marked as finished in the checkpoint. If all the subtask of the operators are finished, the execution of the operator would be skipped after the failover. If not all subtask are finished, the remaining states will be re-assigned, but we assume the produced records for this operator as a whole should not changed.  A more detailed explanation is in [here](https://cwiki.apache.org/confluence/display/FLINK/FLIP-147%3A+Support+Checkpoints+After+Tasks+Finished#FLIP147:SupportCheckpointsAfterTasksFinished-Option3.Allowtaskstofinish&Checkpointsdonotcontainthefinalstatesfromfinishedtasks).

> FLINK-21085: I think the currently solution will undermine what we try to follow in FLINK-21133. In that jira, we try to unify handling of stop-with-savepoint in one place, while the currently approach tends to duplicate checkpoint trigger in many code paths. I would like to suggest to enhance and generalize MultipleInputStreamTask.triggerCheckpointAsync to StreamTask. This way we will have only place to trigger checkpoint.

Sorry I'm not quite sure what the duplication refers to, would you detail this issue~?

> FLINK-21081: Just yet another approach to list/consider/evaluation. May be we could send checkpoint trigger rpc to all running tasks ? If tasks with active inputs receive checkpoint trigger rpc, just bookkeeping it. If all active inputs are EndOfPartitionEvent in the meantime, task will trigger checkpoint themselves. If checkcpoint-trigger is coded in one place, this bookkeeping and lazy-trigger will only be coded once in one place. But this approach may be network-consuming in large job topology.

Personally I would not lean towards this method due to the possible costs for triggering all the tasks, especially for large jobs. 

> I am no sure how finishing handshake will be delivered, so basically, I am also not sure how this handshaking will combine with stop-with-savepoint. But if the handshaking happens at StreamTask level, I think there will be no big deal. Though, still unsure.

I pushed the implementation to this branch: https://github.com/gaoyunhaii/flink/commits/final_cp_6_handshark, it is almost done. Any comments are welcome~


> Assumes that operator C has two upstream A and B. During a checkpoint, A ends with EndOfPartitionEvent while B keeps running. C could receive checkpoint barrier from B and then EndOfPartitionEvent from A. In this case, FLINK-21081 does not help.

We should have implementation this in the above new branch, In the new implementation, when the channel received `EndOfPartitionEvent` would also be viewed as aligned and the checkpoint could be trigger normally.

**Comment by kezhuw** [comment]
@gaoyunhaii Sorry, I propably targets wrong method, I means `StreamTask.triggerCheckpoint` and its overrides. May be not that serious, but I think it would be good to code trigger-checkpoint in one place.

I have taken an quick look at new implementation:
* For handling of `EndOfPartitionEvent`, I think it is much straightforward than old `FinalizeBarrierComplementPrcoessor`.
* For "overtaking buffers for unaligned checkpoint", I think both methods("flushed to network stack" and "processed by downstream") are achievable.
* "No, the waiting happens after all the operators have been closed"
    * Please aware of FLINK-16383 which ignores `notifyCheckpointCompleted` after operator closed. Personally, I think it is problem of `StreamOperator.close` which is documented as "flushing buffered data" but misused(?) by many operators as  "cleanup-resources".
    * What if `StreamOperator.close` generates some new records, assumes that `StreamOperator.close` is the end-of-stream-flush? Will these records be duplicated in recovery ?

**Comment by kezhuw** [comment]
Hi @gaoyunhaii, I have seen FLIP-147 for reference about caveat of end-of-input handling. Basically, I have no more concerns, assumes that end-of-input should only flush buffered/aggregated but not new records. Hope it is documented somewhere in javadoc in future pr for that end-of-input-flushing method.

**Comment by gaoyunhaii** [comment]
@kezhuw Very thanks for the thoughts and suggestions! I'll have a look on how to add the comment~
